com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/BreakpointManager.java

```diff
@@ -31,6 +31,9 @@
     private List<IBreakpoint> breakpoints;
     private HashMap<String, HashMap<String, IBreakpoint>> sourceToBreakpoints;
     private AtomicInteger nextBreakpointId = new AtomicInteger(1);
+    // BreakpointManager is the owner class of the breakpoint to compiled expression map, it will remove
+    // the breakpoint from this map if the breakpoint is removed or its condition is changed
+    private Map<IBreakpoint, Object> breakpointExpressionMap = new HashMap<>();
```

Comment:
Why is the breakpoint manager maintaining the cache map but not the evaluation engine?

Reply:
The cache should be kept by evaluation engine. The BreakpointManger notifies the engine to invalidate cache entries.

Reply:
Because evaluation engine has no information about when the breakpoint is removed or updated.